clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3431,23 +3416,100 @@ static void handleSYCLIntelSchedulerTargetFmaxMhzAttr(Sema &S, Decl *D,
 }
 
 // Handles max_global_work_dim.
-static void handleMaxGlobalWorkDimAttr(Sema &S, Decl *D, const ParsedAttr &A) {
-  if (D->isInvalidDecl())
-    return;
+void Sema::AddSYCLIntelMaxGlobalWorkDimAttr(Decl *D,
+                                            const AttributeCommonInfo &CI,
+                                            Expr *E) {
+  if (!E->isValueDependent()) {
+    // Validate that we have an integer constant expression and then store the
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
 
-  Expr *E = A.getArgAsExpr(0);
+    // This attribute requires a non-negative value.
+    if (ArgVal < 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*non-negative*/ 1;
+      return;
+    }
+    // This attribute must be in the range [0, 3].
+    if (ArgVal > 3) {
+      Diag(E->getBeginLoc(), diag::err_attribute_argument_out_of_range)
+          << CI << 0 << 3 << E->getSourceRange();
+      return;
+    }
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
+    if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+      // If the other attribute argument is instantiation dependent, we won't
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      if (const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue())) {
+        if (ArgVal != DeclExpr->getResultAsAPSInt()) {
+          Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+          Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+        }
+        // If there is no mismatch, drop any duplicate attributes.
+        return;
+      }
+    }
+    // If the declaration has a SYCLIntelMaxWorkGroupSizeAttr or
+    // ReqdWorkGroupSizeAttr, check to see if they hold equal values
+    // (1, 1, 1) in case the value of SYCLIntelMaxGlobalWorkDimAttr
+    // equals to 0.
+    auto Check = [this, &CI](const auto *A) {
+      Optional<llvm::APSInt> XDimVal = A->getXDimVal(Context);
+      Optional<llvm::APSInt> YDimVal = A->getYDimVal(Context);
+      Optional<llvm::APSInt> ZDimVal = A->getZDimVal(Context);
+      if (*XDimVal != 1 || *YDimVal != 1 || *ZDimVal != 1) {
```

Comment:
added test case for this.

Reply:
Perhaps we need a different test -- I was expecting to see changes here when X, Y, or Z are dependent values, so the `Optional` would hold `None`. Was that not the case for my example?

Reply:
we are doing wrong here. Are you concerned the test cases below?
```
 error: no matching function for call to 'func1'
  func1<3.f>();
  ^~~~~~~~~~
re.cpp:4:40: note: candidate template ignored: invalid explicitly-specified argument for template parameter 'N'
[[intel::max_global_work_dim(0)]] void func1();
                                       ^
1 error generated.

template <int N>
[[intel::reqd_work_group_size(N, N, N)]] void func1();
template <int N>
[[intel::max_global_work_dim(0)]] void func1();

int main() {
  func1<3.f>();
  return 0;
}
````

Reply:
I'm not worried about that case -- the template requires a nontype template parameter of type `int` and you're specifying a `float`, so that's why you get the "no matching function call" error. You get the same with: https://godbolt.org/z/xW4znE5a4

I don't think you need to instantiate the function to trigger the issue I'm worried about:
```
template <int N>
[[intel::max_work_group_size(N, N, N)]] void func(); // #1
template <int N>
[[intel::max_global_work_dim(0)]] void func(); // #2
```
At (1) we add the `max_work_group_size` attribute to `func()` but the attribute arguments are value dependent. At (2) when adding the `max_global_work_dim` attribute, we are checking whether the `max_work_group_size` arguments are correct, but we still can't do that because there's still not an instantiation at that point, so the `Optional`s should come back as `None`.

Then we would want two tests: `func<1>()` (which should work) and `func<2>()` (which should have an error because th `max_work_group_size` args are `2, 2, 2`).